### DISCUSSION TITLE
CSS vars for shades of primaryColor

### DISCUSSION THREAD

**Discussion by aviemet** [mantinedev/mantine#5117] [Q&A]
Any ideas how I might go about using shades of the primary color? I'd like to be able to do something along the lines of `var(--mantine-color-primary-2)` from within CSS modules. I can't seem to figure out how to set this up in an intuitive way so that the css variable would change and update everything if the primary color changed.

**Comment by rtivital** [comment]
You can add any additional CSS variables based on your theme object with CSS variables resolver – https://mantine.dev/styles/css-variables/#css-variables-resolver

**Comment by aviemet** [comment]
Thanks, I just want to post my solution and discuss it a tiny bit. Here's how I've gone about this:

In order to generate the desired css variable names in a `vars` object, I added a `primary` key with an irrelevant value to the `themeToVars` call. I found that if I added this to the theme object itself, `theme.colors.primary` still behaved as a string rather than an object.

```typescript
export const vars = themeToVars(mergeMantineTheme(theme, {
  colors: {
    primary: DEFAULT_THEME.colors[defaultColor],
  },
}))
```

Then I manually set the values for primary color variants to the current primary color.

```tsx
const UiFrameworkProvider = ({ children }: { children: React.ReactNode }) => {
  const { primaryColor, setPrimaryColor } = useLayoutStore()

  const theme = createTheme({ ...themeObject, primaryColor })

  const cssVariablesResolver: CSSVariablesResolver = (resolverTheme) => {
    const variables: CSSVariables = {}
    const dark: CSSVariables = {}
    const light: CSSVariables = {}

    Object.entries(vars.colors[primaryColor]).forEach(([key, val]) => {
      if(key.match(/[0-9]/)) {
        variables[`--mantine-color-primary-${toKebabCase(key)}`] = val
      } else {
        dark[`--mantine-color-primary-${toKebabCase(key)}`] = val
        light[`--mantine-color-primary-${toKebabCase(key)}`] = val
      }
    })

    return { variables, dark, light }
  }

  return (
    <MantineProvider
      theme={ theme }
      defaultColorScheme="auto"
      cssVariablesResolver={ cssVariablesResolver }
    >
      { children }
    </MantineProvider>
  )
}
```

So now there are CSS variables set on the root element like `--mantine-color-primary-4: var(--mantine-color-orange-4);`, etc, and I can access it in `css.ts` files using `vars.colors.primary[4]`.

This works, and is the best I've come up with so far, but it has the drawback that now the value of `vars.colors.primary` is a `Colors` object rather than a `string`, which means I'm out of sync with the docs. I've also had a very hard time figuring out how to do the typescript declaration merging for this, so I'm actually just directly overriding the type using `as`.

Before v7, we could do `theme.colors[theme.primaryColor][4]` to access the variants in emotion. As far as I can tell, doing this sort of hacky thing I did is the best way to get that functionality. It does make me wonder if there's any will to have this built in to the framework? Since the only css variable dealing with the primary key is `--mantine-color-primary`, it wouldn't be a breaking change to the style system. The only breaking change would be that the `colors.primary` key returned from `themeToVars` wouldn't be a string.

Also, since this is a Q&A post and I've already made it super long, I would sure appreciate any suggestions to improve this. Already mentioned that I still haven't figured out how to properly override the `themeToVars` return value type, but I also suspect that adding the same values to the `dark` and the `light` variants is not quite right.
